api/src/main/java/net/kyori/adventure/text/width/PixelWidthSourceImpl.java

```diff
@@ -0,0 +1,368 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.text.width;
+
+import net.kyori.adventure.text.BlockNBTComponent;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.EntityNBTComponent;
+import net.kyori.adventure.text.KeybindComponent;
+import net.kyori.adventure.text.NBTComponent;
+import net.kyori.adventure.text.ScoreComponent;
+import net.kyori.adventure.text.SelectorComponent;
+import net.kyori.adventure.text.StorageNBTComponent;
+import net.kyori.adventure.text.TextComponent;
+import net.kyori.adventure.text.TranslatableComponent;
+import net.kyori.adventure.text.format.Style;
+import net.kyori.adventure.text.format.TextDecoration;
+import net.kyori.adventure.text.renderer.TranslatableComponentRenderer;
+import net.kyori.adventure.translation.GlobalTranslator;
+import net.kyori.adventure.util.CharToIntFunction;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import static net.kyori.adventure.text.Component.text;
+
+/**
+ * An implementation of the pixel width source which handles serialization with a set of functions
+ * for resolving all components except {@link TextComponent}s.
+ *
+ * @param <CX> a context (player, server, locale)
+ * @since 4.5.0
+ */
+
+class PixelWidthSourceImpl<CX> implements PixelWidthSource<CX> {
+
+  private final Set<ComponentResolver<KeybindComponent, CX>> keybindResolvers = new HashSet<>();
+  private final Set<ComponentResolver<SelectorComponent, CX>> selectorResolvers = new HashSet<>();
+  private final Set<ComponentResolver<ScoreComponent, CX>> scoreResolvers = new HashSet<>();
+  private final Set<ComponentResolver<BlockNBTComponent, CX>> blockNBTResolvers = new HashSet<>();
+  private final Set<ComponentResolver<EntityNBTComponent, CX>> entityNBTResolvers = new HashSet<>();
+  private final Set<ComponentResolver<StorageNBTComponent, CX>> storageNBTResolvers = new HashSet<>();
+  private final Set<ComponentResolver<TranslatableComponent, CX>> translatableResolvers = new HashSet<>();
+
+  private final PixelWidthRenderer renderer = new PixelWidthRenderer();
+
+  private final Function<CX, CharToIntFunction<Style>> function;
+  private final Function<CX, Locale> localeFunction;
+
+  /**
+   * Get the pixel width of the given character in the Minecraft font, excluding the space between
+   * characters and drop shadow. Handles alphanumerics and most common english punctuation.
+   */
+  static final CharToIntFunction<Style> DEFAULT_FONT_WIDTH = (c, style) -> {
```

Comment:
This should be something generated by `adventure-vanilla-data` or otherwise directly by the client, not a magic block of code that's been copied around.

Reply:
I think the API should provide a default, otherwise it's going to be too complicated for the average API user to use it.

Reply:
I never said it shouldn't be part of the API -- this should just be something reproducible based on data for any font, not magic.